packages/backend/server/src/plugins/copilot/workflow/node.ts

```diff
@@ -1,21 +1,70 @@
-import { ChatPrompt, PromptService } from '../prompt';
-import { CopilotProviderService } from '../providers';
-import { CopilotAllProvider, CopilotChatOptions } from '../types';
-import {
+import path, { dirname } from 'node:path';
+import { fileURLToPath } from 'node:url';
+
+import { Logger } from '@nestjs/common';
+import Piscina from 'piscina';
+
+import { CopilotChatOptions } from '../types';
+import { getWorkflowExecutor, WorkflowExecutor } from './executor';
+import type {
   NodeData,
+  WorkflowGraph,
   WorkflowNodeState,
-  WorkflowNodeType,
   WorkflowResult,
-  WorkflowResultType,
 } from './types';
+import { WorkflowNodeType, WorkflowResultType } from './types';
 
 export class WorkflowNode {
+  private readonly logger = new Logger(WorkflowNode.name);
   private readonly edges: WorkflowNode[] = [];
   private readonly parents: WorkflowNode[] = [];
-  private prompt: ChatPrompt | null = null;
-  private provider: CopilotAllProvider | null = null;
-
-  constructor(private readonly data: NodeData) {}
+  private readonly executor: WorkflowExecutor | null = null;
+  private readonly condition:
+    | ((params: WorkflowNodeState) => Promise<any>)
+    | null = null;
+
+  constructor(
+    graph: WorkflowGraph,
+    private readonly data: NodeData
+  ) {
+    if (data.nodeType === WorkflowNodeType.Basic) {
+      this.executor = getWorkflowExecutor(data.type);
+    } else if (data.nodeType === WorkflowNodeType.Decision) {
+      // prepare decision condition, reused in each run
+      const iife = `(${data.condition})(nodeIds, params)`;
+      // only eval the condition in worker if graph has been modified
+      if (graph.modified) {
+        const worker = new Piscina({
+          filename: path.resolve(
+            dirname(fileURLToPath(import.meta.url)),
+            'worker.mjs'
+          ),
+          minThreads: 2,
+          // empty envs from parent process
+          env: {},
+          argv: [],
+          execArgv: [],
+        });
+        this.condition = (params: WorkflowNodeState) =>
+          worker.run({
+            iife,
+            nodeIds: this.edges.map(node => node.id),
+            params,
+          });
+      } else {
+        const func = new Function(
```

Comment:
make data.condition type `fn | string`? that would be more safe and efficient cause don't trust `fn.toString()` much